dragonfly/implementations/psyllid_provider.py

```diff
@@ -0,0 +1,279 @@
+class MultiPsyllidProvider(core.Provider):
+    '''
+    Provider for direct communication with up to 3 Psyllid instances with a single stream each 
+    '''
+    def __init__(self,
+                 queue_a='channel_a_psyllid',
+                 queue_b = 'channel_b_psyllid',
+                 queue_c = 'channel_c_psyllid',
+                 set_condition_list = [],
+                 **kwargs):
+
+        core.Provider.__init__(self, **kwargs)        
+        self.queue_dict = {'a':queue_a, 'b':queue_b, 'c':queue_c}
+        self._set_condition_list = set_condition_list
+        self.status_dict = {'a':None, 'b':None, 'c':None}
+        self.status_value_dict = {'a':None, 'b':None, 'c':None}
+        self.channel_dict = {'a': 'ch0', 'b': 'ch1', 'c': 'ch2'}
+        self.freq_dict = {'a': None, 'b': None, 'c': None}
+        self.mode_dict = {'a':None, 'b':None, 'c':None}
+
+
+    def _finish_configure(self):
+        for channel in self.channel_dict.keys():
+            if self.request_status(channel)!=None:
+                self.mode_dict[channel]='streaming'
+
+                try: 
+                    if self.set_central_frequency(channel=channel, cf=800e6)==False:
+                       self.mode_dict[channel]='triggered'
+                       if self.set_central_frequency(channel=channel, cf=800e6)==False:
+                            self.mode_dict[channel]=None
+                            raise core.exceptions.DriplineGenericDAQError("Stream writer of Psyllid instance for channel {} is not configured correctly".format(channel))
+
+                    if self.get_number_of_streams(channel)!=1:
+                        self.mode_dict[channel]=None
+                        raise core.exceptions.DriplineGenericDAQError("Streams of Psyllid instance for channel {} are not configured correctly".format(channel))
+
+                except core.exceptions.DriplineGenericDAQError:
+                    logger.warning('No matching Psyllid instance for channel {} present'.format(channel))
+                except Exception as e:
+                    logger.error('Unexpected error during startup')
+                    raise 
+
+        # Summary after startup
+        logger.info('Status of channels: {}'.format(self.status_value_dict))
+        logger.info('Set central frequencies: {}'.format(self.freq_dict))
+        logger.info('Streaming or triggered mode: {}'.format(self.mode_dict))
+
+
+    def request_status(self, channel):
+        logger.info('Checking Psyllid status of channel {}'.format(channel))
+        try:
+            result = self.provider.get(self.queue_dict[channel]+'.daq-status', timeout=10)
+        except core.exceptions.DriplineError:
+            logger.warning('Psyllid instance for channel {} is not running or returned error'.format(channel))
+            self.status_dict[channel]=None
+            self.status_value_dict[channel]=None
+            return self.status_value_dict[channel]
```

Comment:
Sorry, I know this form exists in several places but I think I understand the logic better now. Please correct the following:

1) attempt to send a Request to the psyllid service asking for the `daq-status`
2.1) If the a normal Reply is received, do the stuff in the `else` block. (A couple of questions here: i) is there a reason to put it in an else block rather than as direct statements at one less indentation level? I assume not because every exception you catch either returns or raises a new exception, but this perhaps more explicit so I'm not suggesting changing it. ii) Is the purposes of self.status_dict and status_value_dict simply to cache the state of psyllid services, and if so is there ever a case where you care about those states when you use the cached value rather than polling again? My guess is that you don't actually need to cache the value (it is more consistent dripline design not to do so), and that at most you'd want to save those states in a local variable for things like getting/checking the state of all the required psyllid instances before starting a run that talks to several. But I haven't gone through the logic of the full file so maybe I'm missing something
2.2) If the reply failed you set the cache values to None and then return that. I wonder if this is really the desired behavior. Is it ever the case that this happens and it is not an indication that the user needs to go start/restart psyllid or otherwise check it for errors? I assume we want to pass something back to the user so that they know that. Either by re-raising the existing exception, or raising a new one with more information if we have it. If there are a mix of cases we could also return a warning returncode (value less than 100) which is allowed to include a payload with useful data for the parent process to continue executing, but also indicates that something went wrong-ish.